<a href="https://colab.research.google.com/github/haroharsh/KeyboardShortcuts/blob/main/R%26FDocs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
%pip install tesseract

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.6/45.6 MB 18.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for tesseract: filename=tesseract-0.1.3-py3-none-any.whl size=45562552 sha256=4b2c271828cca26d9e847c8b5f836d44296726f020e81637d0325291029d2478
  Stored in directory: /root/.cache/pip/wheels/13/1f/8e/2d6c0e358fd6d01ca80ecd9185a374bcda35879f4fec727242
Successfully built tesseract


In [4]:
import pandas as pd

# Replace 'your_file.csv' with the actual name of your uploaded file
df = pd.read_csv('SSL_dataset.csv')

# Display the first few rows of the DataFrame to verify
display(df.head())

RSAEncryption'  \
safehorizon.org       R3                                                 15/04/2023 17:26 14/07/2023 17:26 3.750000e+41 <OpenSSL.crypto.PKey object at 0x000001AA0CA00D60>  b'sha256WithRSAEncryption'   
podcastindex.org      Cloudflare Inc ECC CA-3                            17/03/2023 0:00  16/03/2024 23:59 1.390000e+37 <OpenSSL.crypto.PKey object at 0x000001AA0CA07E50>        b'ecdsa-with-SHA256'   
sni.cloudflaressl.com Cloudflare Inc ECC CA-3                            19/06/2022 0:00  18/06/2023 23:59 1.620000e+36 <OpenSSL.crypto.PKey object at 0x000001AA0CA077F0>        b'ecdsa-with-SHA256'   
*.mucem.org           Sectigo RSA Organization Validation Secure Serv... 24/04/2023 0:00  11/05/2024 23:59 2.610000e+38 <OpenSSL.crypto.PKey object at 0x000001AA0C9D9850>  b'sha256WithRSAEncryption'   
*.hudsonriverblue.com R3                                                 8/04/2023 16:46  7/07/2023 16:46  3.920000e+41 <OpenSSL.crypto.PKey object at 0x000001AA0C328760>  b'sha256WithRSAEncryption'   

                                                                                                                                                                            US  \
safehorizon.org       R3                                                 15/04/2023 17:26 14/07/2023 17:26 3.750000e+41 <OpenSSL.crypto.PKey object at 0x000001AA0CA00D60>  US   
podcastindex.org      Cloudflare Inc ECC CA-3                            17/03/2023 0:00  16/03/2024 23:59 1.390000e+37 <OpenSSL.crypto.PKey object at 0x000001AA0CA07E50>  US   
sni.cloudflaressl.com Cloudflare Inc ECC CA-3                            19/06/2022 0:00  18/06/2023 23:59 1.620000e+36 <OpenSSL.crypto.PKey object at 0x000001AA0CA077F0>  US   
*.mucem.org           Sectigo RSA Organization Validation Secure Serv... 24/04/2023 0:00  11/05/2024 23:59 2.610000e+38 <OpenSSL.crypto.PKey object at 0x000001AA0C9D9850>  GB   
*.hudsonriverblue.com R3                                                 8/04/2023 16:46  7/07/2023 16:46  3.920000e+41 <OpenSSL.crypto.PKey object at 0x000001AA0C328760>  US   

                                                                                                                                                                                    Unnamed: 2  \
safehorizon.org       R3                                                 15/04/2023 17:26 14/07/2023 17:26 3.750000e+41 <OpenSSL.crypto.PKey object at 0x000001AA0CA00D60>                 NaN   
podcastindex.org      Cloudflare Inc ECC CA-3                            17/03/2023 0:00  16/03/2024 23:59 1.390000e+37 <OpenSSL.crypto.PKey object at 0x000001AA0CA07E50>                 NaN   
sni.cloudflaressl.com Cloudflare Inc ECC CA-3                            19/06/2022 0:00  18/06/2023 23:59 1.620000e+36 <OpenSSL.crypto.PKey object at 0x000001AA0CA077F0>                 NaN   
*.mucem.org           Sectigo RSA Organization Validation Secure Serv... 24/04/2023 0:00  11/05/2024 23:59 2.610000e+38 <OpenSSL.crypto.PKey object at 0x000001AA0C9D9850>  Greater Manchester   
*.hudsonriverblue.com R3                                                 8/04/2023 16:46  7/07/2023 16:46  3.920000e+41 <OpenSSL.crypto.PKey object at 0x000001AA0C328760>                 NaN   

                                                                                                                                                                           Unnamed: 3  \
safehorizon.org       R3                                                 15/04/2023 17:26 14/07/2023 17:26 3.750000e+41 <OpenSSL.crypto.PKey object at 0x000001AA0CA00D60>        NaN   
podcastindex.org      Cloudflare Inc ECC CA-3                            17/03/2023 0:00  16/03/2024 23:59 1.390000e+37 <OpenSSL.crypto.PKey object at 0x000001AA0CA07E50>        NaN   
sni.cloudflaressl.com Cloudflare Inc ECC CA-3                            19/06/2022 0:00  18/06/2023 23:59 1.620000e+36 <OpenSSL.crypto.PKey object at 0x000001AA0CA077F0>        NaN 

## Exploring Different Classification Algorithms

### Subtask:
Train and evaluate a Random Forest Classifier to compare its performance with the Logistic Regression model.

In [5]:
# Check data types
print("Data Types:")
print(df.dtypes)

# Get summary statistics for numerical columns
print("\nSummary Statistics:")
display(df.describe())

# Check for missing values
print("\nMissing Values:")
print(df.isnull().sum())

Data Types:
RSAEncryption'     object
US                 object
Unnamed: 2         object
Unnamed: 3         object
Let's Encrypt      object
Unnamed: 5         object
0.0               float64
dtype: object

Summary Statistics:


0.0
count  7484.000000
mean      0.019642
std       0.138776
min       0.000000
25%       0.000000
50%       0.000000
75%       0.000000
max       1.000000


Missing Values:
RSAEncryption'       0
US                  37
Unnamed: 2        6280
Unnamed: 3        6263
Let's Encrypt       37
Unnamed: 5        6636
0.0                  0
dtype: int64


In [2]:
%pip install pytesseract transformers scikit-learn

# Task
Build a machine learning model to classify data from a CSV file.

## Data preprocessing

### Subtask:
Prepare the data for modeling, including handling missing values, encoding categorical features, and splitting the data into training and testing sets.


**Reasoning**:
Drop the specified unnamed columns, drop rows with any remaining missing values, and then proceed with one-hot encoding for the categorical columns and splitting the data.



In [6]:
from sklearn.model_selection import train_test_split
import pandas as pd

# Drop the specified unnamed columns
df = df.drop(columns=['Unnamed: 2', 'Unnamed: 3', 'Unnamed: 5'])

# Drop rows with missing values
df.dropna(inplace=True)

# Convert categorical columns to string and apply one-hot encoding
df['RSAEncryption\''] = df['RSAEncryption\''].astype(str)
df = pd.get_dummies(df, columns=['RSAEncryption\''], prefix='RSAEncryption')

df['US'] = df['US'].astype(str)
df = pd.get_dummies(df, columns=['US'], prefix='Country')

df['Let\'s Encrypt'] = df['Let\'s Encrypt'].astype(str)
df = pd.get_dummies(df, columns=['Let\'s Encrypt'], prefix='Issuer')

# Separate features (X) and target (y)
X = df.drop(columns=['0.0'])
y = df['0.0']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (5956, 141)
Shape of X_test: (1489, 141)
Shape of y_train: (5956,)
Shape of y_test: (1489,)


## Model selection

### Subtask:
Choose an appropriate machine learning model based on the problem you are trying to solve (e.g., classification, regression).


**Reasoning**:
Based on the problem being a binary classification task, Logistic Regression and Random Forest are appropriate choices. Logistic Regression provides a simple yet effective linear model, while Random Forest is a more complex ensemble method that can capture non-linear relationships. I will choose Logistic Regression first as it's generally a good baseline for classification problems.



In [7]:
from sklearn.linear_model import LogisticRegression

# Initialize the Logistic Regression model
model = LogisticRegression(random_state=42, solver='liblinear')

## Model training

### Subtask:
Train the selected Logistic Regression model using the training data.


**Reasoning**:
Fit the initialized Logistic Regression model to the training data.



In [8]:
# Fit the model to the training data
model.fit(X_train, y_train)

LogisticRegression(random_state=42, solver='liblinear')

## Model evaluation

### Subtask:
Evaluate the performance of the trained Logistic Regression model using the testing data and appropriate metrics.


**Reasoning**:
Evaluate the trained Logistic Regression model using the testing data and calculate relevant classification metrics.



In [9]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Predict the target variable for the test set
y_pred = model.predict(X_test)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Print the evaluation metrics
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1 Score: {f1:.4f}")

Accuracy: 0.9953
Precision: 1.0000
Recall: 0.6500
F1 Score: 0.7879


## Summary:

### Data Analysis Key Findings

*   The dataset was preprocessed by dropping 'Unnamed: 2', 'Unnamed: 3', and 'Unnamed: 5' columns, handling missing values, and applying one-hot encoding to the 'RSAEncryption', 'US', and 'Let\'s Encrypt' columns.
*   The data was split into training and testing sets, with 80% for training and 20% for testing.
*   A Logistic Regression model was selected and trained on the preprocessed training data.
*   The trained Logistic Regression model achieved the following performance metrics on the test set:
    *   Accuracy: 0.9953
    *   Precision: 1.0000
    *   Recall: 0.6500
    *   F1 Score: 0.7879

### Insights or Next Steps

*   The high precision (1.0000) indicates that when the model predicts the positive class, it is always correct. However, the lower recall (0.6500) suggests the model misses a significant portion of the actual positive cases.
*   Investigate techniques to improve the recall of the model, such as exploring different classification algorithms, adjusting class weights, or performing further feature engineering.
